models/glif_lif_asc.h

```diff
@@ -0,0 +1,285 @@
+
+  void handle( nest::SpikeEvent& );
+  void handle( nest::CurrentEvent& );
+  void handle( nest::DataLoggingRequest& );
+
+  nest::port handles_test_event( nest::SpikeEvent&, nest::port );
+  nest::port handles_test_event( nest::CurrentEvent&, nest::port );
+  nest::port handles_test_event( nest::DataLoggingRequest&, nest::port );
+
+  bool is_off_grid() const // uses off_grid events
+  {
+    return true;
+  }
+
+  void get_status( DictionaryDatum& ) const;
+  void set_status( const DictionaryDatum& );
+
+private:
+  //! Reset parameters and state of neuron.
+
+  //! Reset state of neuron.
+  void init_state_( const Node& proto );
+
+  //! Reset internal buffers of neuron.
+  void init_buffers_();
+
+  //! Initialize auxiliary quantities, leave parameters and state untouched.
+  void calibrate();
+
+  //! Take neuron through given time interval
+  void update( nest::Time const&, const long, const long );
+
+  // The next two classes need to be friends to access the State_ class/member
+  friend class nest::RecordablesMap< glif_lif_asc >;
+  friend class nest::UniversalDataLogger< glif_lif_asc >;
+
+
+  struct Parameters_
+  {
+    double V_th_;    // A constant spiking threshold in mV
+    double G_;       // membrane conductance in nS
+    double E_L_;     // resting potential in mV
+    double C_m_;     // capacitance in pF
+    double t_ref_;   // refractory time in ms
+    double V_reset_; // Membrane voltage following spike in mV
+
+    std::vector< double > asc_init_; // initial values of ASCurrents_ in pA
+    std::vector< double > k_;        // predefined time scale in 1/ms
+    std::vector< double > asc_amps_; // in pA
+    std::vector< double > r_;        // coefficient
+    std::string V_dynamics_method_;  // voltage dynamic methods
+
+    Parameters_();
+
+    void get( DictionaryDatum& ) const;
+    void set( const DictionaryDatum& );
+  };
+
+
+  struct State_
+  {
+    double V_m_;                       // membrane potential in mV
+    std::vector< double > ASCurrents_; // after-spike currents in pA
+    double ASCurrents_sum_;            // in sum of after-spike currents in pA
+
+    double I_; // external current in pA
+
+    State_();
+
+    void get( DictionaryDatum& ) const;
+    void set( const DictionaryDatum&, const Parameters_& );
+  };
+
+
+  struct Buffers_
+  {
+    Buffers_( glif_lif_asc& );
+    Buffers_( const Buffers_&, glif_lif_asc& );
+
+    nest::RingBuffer spikes_; //!< Buffer incoming spikes through delay, as sum
+    nest::RingBuffer currents_; //!< Buffer incoming currents through delay,
+
+    //! Logger for all analog data
+    nest::UniversalDataLogger< glif_lif_asc > logger_;
+  };
+
+  struct Variables_
+  {
+    double t_ref_remaining_; // counter during refractory period in ms
+    double t_ref_total_;     // total time of refractory period in ms
+    int method_; // voltage dynamics solver method flag: 0-linear forward euler;
+                 // 1-linear exact
```

Comment:
Since the linear exact method is available, is there any reason to include forward Euler? I think the linear exact method might be improved even more by integrating all currents using exact integration; see `mat2_psc_exp` for an example and the appendix of [Heiberg et al (2018)](http://dx.doi.org/10.1007/s10827-018-0693-9) for details.

If you want a `method_` switch, I would suggest to use an `enum` instead of a plain `int`.

Reply:
<method_> has removed. The models are using linear exact solutions.